File: fastapi_backend/lib/aws_s3.py
import boto3
from botocore.exceptions import ClientError, NoCredentialsError
import os
from typing import Dict, Any, Optional, Tuple
from dataclasses import dataclass
import logging
import time
from dotenv import load_dotenv

# Load environment variables from .env file
load_dotenv()

logger = logging.getLogger(__name__)

@dataclass
class UploadProgress:
    loaded: int
    total: int
    percentage: int

@dataclass
class S3UploadResult:
    key: str
    location: str
    bucket: str
    etag: str

class S3MultipartUpload:
    def __init__(self, bucket: Optional[str] = None, key_prefix: str = "videos/"):
        self.bucket = bucket or os.getenv("AWS_S3_BUCKET", "")
        self.key_prefix = key_prefix
        
        if not self.bucket:
            raise ValueError("AWS S3 bucket name is required. Set AWS_S3_BUCKET environment variable.")
        
        # Initialize S3 client
        self.s3_client = boto3.client(
            's3',
            region_name=os.getenv("AWS_REGION", "us-east-1"),
            aws_access_key_id=os.getenv("AWS_ACCESS_KEY_ID"),
            aws_secret_access_key=os.getenv("AWS_SECRET_ACCESS_KEY")
        )
    
    async def upload_file(self, file_data: bytes, file_name: str, content_type: str) -> S3UploadResult:
        """Upload a file to S3"""
        timestamp = int(time.time() * 1000)
        sanitized_name = "".join(c if c.isalnum() or c in ".-_" else "_" for c in file_name)
        key = f"{self.key_prefix}{timestamp}_{sanitized_name}"
        
        try:
            response = self.s3_client.put_object(
                Bucket=self.bucket,
                Key=key,
                Body=file_data,
                ContentType=content_type,
                Metadata={
                    'originalName': file_name,
                    'fileSize': str(len(file_data)),
                    'uploadTimestamp': str(timestamp)
                }
            )
            
            return S3UploadResult(
                key=key,
                location=f"https://{self.bucket}.s3.amazonaws.com/{key}",
                bucket=self.bucket,
                etag=response['ETag'].strip('"')
            )
        except Exception as e:
            logger.error(f"S3 upload failed: {e}")
            raise Exception(f"Failed to upload file to S3: {str(e)}")
    
    async def get_presigned_upload_url(self, file_name: str, file_type: str, file_size: int) -> Dict[str, Any]:
        """Generate a presigned URL for direct client-side upload"""
        timestamp = int(time.time() * 1000)
        sanitized_name = "".join(c if c.isalnum() or c in ".-_" else "_" for c in file_name)
        key = f"{self.key_prefix}{timestamp}_{sanitized_name}"
        
        try:
            logger.info(f"Generating presigned PUT URL for: {file_name}, type: {file_type}, size: {file_size}")
            
            # Generate presigned URL for PUT operation
            upload_url = self.s3_client.generate_presigned_url(
                'put_object',
                Params={
                    'Bucket': self.bucket,
                    'Key': key,
                    'ContentType': file_type,
                    'Metadata': {
                        'original-name': file_name,
                        'file-size': str(file_size),
                        'upload-timestamp': str(timestamp)
                    }
                },
                ExpiresIn=3600  # 1 hour
            )
            
            logger.info(f"Generated presigned PUT URL successfully for key: {key}")
            
            return {
                'uploadUrl': upload_url,
                'key': key
            }
        except Exception as e:
            logger.error(f"Failed to generate presigned PUT URL: {e}")
            raise Exception(f"Failed to generate upload URL: {str(e)}")
    
    async def delete_file(self, key: str) -> None:
        """Delete a file from S3"""
        try:
            self.s3_client.delete_object(Bucket=self.bucket, Key=key)
        except Exception as e:
            logger.error(f"Failed to delete file from S3: {e}")
            raise Exception(f"Failed to delete file: {str(e)}")
    
    async def get_file_info(self, key: str) -> Dict[str, Any]:
        """Get file metadata and check if it exists"""
        try:
            response = self.s3_client.head_object(Bucket=self.bucket, Key=key)
            return {
                'exists': True,
                'size': response.get('ContentLength'),
                'lastModified': response.get('LastModified'),
                'contentType': response.get('ContentType')
            }
        except ClientError as e:
            if e.response['Error']['Code'] == '404':
                return {'exists': False}
            raise Exception(f"Failed to get file info: {str(e)}")
    
    async def download_file(self, key: str) -> Dict[str, Any]:
        """Download a file from S3"""
        try:
            response = self.s3_client.get_object(Bucket=self.bucket, Key=key)
            return {
                'buffer': response['Body'].read(),
                'contentType': response.get('ContentType'),
                'contentLength': response.get('ContentLength')
            }
        except Exception as e:
            logger.error(f"Failed to download file from S3: {e}")
            raise Exception(f"Failed to download file: {str(e)}")

def validate_aws_config() -> Dict[str, Any]:
    """Validate AWS configuration"""
    required_vars = [
        'AWS_ACCESS_KEY_ID',
        'AWS_SECRET_ACCESS_KEY', 
        'AWS_REGION',
        'AWS_S3_BUCKET'
    ]
    
    missing = [var for var in required_vars if not os.getenv(var)]
    
    return {
        'isValid': len(missing) == 0,
        'missing': missing
    }

# Create default instances
s3_upload = S3MultipartUpload()
s3_downloader = S3MultipartUpload()  # For downloading files 

File: lib/aws-s3.ts
import { S3Client } from "@aws-sdk/client-s3";
import { Upload } from "@aws-sdk/lib-storage";

// AWS S3 Configuration
const s3Client = new S3Client({
  region: process.env.AWS_REGION || "us-east-1",
  credentials: {
    accessKeyId: process.env.AWS_ACCESS_KEY_ID!,
    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY!,
  },
});

export interface UploadProgress {
  loaded: number;
  total: number;
  percentage: number;
}

export interface S3UploadResult {
  key: string;
  location: string;
  bucket: string;
  etag: string;
}

export class S3MultipartUpload {
  private bucket: string;
  private keyPrefix: string;

  constructor(bucket?: string, keyPrefix: string = "videos/") {
    this.bucket = bucket || process.env.AWS_S3_BUCKET || "";
    this.keyPrefix = keyPrefix;
    
    if (!this.bucket) {
      throw new Error("AWS S3 bucket name is required. Set AWS_S3_BUCKET environment variable.");
    }
  }

  /**
   * Upload a file to S3 using multipart upload with progress tracking
   */
  async uploadFile(
    file: File,
    onProgress?: (progress: UploadProgress) => void
  ): Promise<S3UploadResult> {
    // Generate unique key for the file
    const timestamp = Date.now();
    const sanitizedName = file.name.replace(/[^a-zA-Z0-9.-]/g, "_");
    const key = `${this.keyPrefix}${timestamp}_${sanitizedName}`;

    try {
      // Convert File to Buffer for server-side upload
      const arrayBuffer = await file.arrayBuffer();
      const buffer = Buffer.from(arrayBuffer);

      // Create multipart upload
      const upload = new Upload({
        client: s3Client,
        params: {
          Bucket: this.bucket,
          Key: key,
          Body: buffer,
          ContentType: file.type,
          Metadata: {
            originalName: file.name,
            fileSize: file.size.toString(),
            uploadTimestamp: timestamp.toString(),
          },
        },
        // Configure multipart upload settings
        partSize: 10 * 1024 * 1024, // 10 MB parts
        leavePartsOnError: false, // Clean up on failure
      });

      // Track upload progress
      if (onProgress) {
        upload.on("httpUploadProgress", (progress) => {
          const loaded = progress.loaded || 0;
          const total = progress.total || file.size;
          const percentage = Math.round((loaded / total) * 100);
          
          onProgress({
            loaded,
            total,
            percentage,
          });
        });
      }

      // Execute upload
      const result = await upload.done();

      return {
        key: result.Key!,
        location: result.Location!,
        bucket: result.Bucket!,
        etag: result.ETag!,
      };
    } catch (error) {
      console.error("S3 upload failed:", error);
      throw new Error(`Failed to upload file to S3: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Generate a presigned URL for direct client-side upload using PUT method
   */
  async getPresignedUploadUrl(
    fileName: string,
    fileType: string,
    fileSize: number
  ): Promise<{
    uploadUrl: string;
    key: string;
    fields?: Record<string, string>;
  }> {
    const timestamp = Date.now();
    const sanitizedName = fileName.replace(/[^a-zA-Z0-9.-]/g, "_");
    const key = `${this.keyPrefix}${timestamp}_${sanitizedName}`;

    try {
      const { getSignedUrl } = await import("@aws-sdk/s3-request-presigner");
      const { PutObjectCommand } = await import("@aws-sdk/client-s3");
      
      console.log(`Generating presigned PUT URL for: ${fileName}, type: ${fileType}, size: ${fileSize}`);
      
      // Create a PUT object command
      const command = new PutObjectCommand({
        Bucket: this.bucket,
        Key: key,
        ContentType: fileType,
        Metadata: {
          'original-name': fileName,
          'file-size': fileSize.toString(),
          'upload-timestamp': timestamp.toString(),
        },
      });

      // Generate presigned URL for PUT operation
      const uploadUrl = await getSignedUrl(s3Client, command, {
        expiresIn: 3600, // 1 hour
      });

      console.log(`Generated presigned PUT URL successfully for key: ${key}`);

      return {
        uploadUrl,
        key,
        // No fields needed for PUT method
      };
    } catch (error) {
      console.error("Failed to generate presigned PUT URL:", error);
      console.error("Error details:", JSON.stringify(error, null, 2));
      throw new Error(`Failed to generate upload URL: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Delete a file from S3
   */
  async deleteFile(key: string): Promise<void> {
    try {
      const { DeleteObjectCommand } = await import("@aws-sdk/client-s3");
      
      await s3Client.send(new DeleteObjectCommand({
        Bucket: this.bucket,
        Key: key,
      }));
    } catch (error) {
      console.error("Failed to delete file from S3:", error);
      throw new Error(`Failed to delete file: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Get file metadata and check if it exists
   */
  async getFileInfo(key: string): Promise<{
    exists: boolean;
    size?: number;
    lastModified?: Date;
    contentType?: string;
  }> {
    try {
      const { HeadObjectCommand } = await import("@aws-sdk/client-s3");
      
      const result = await s3Client.send(new HeadObjectCommand({
        Bucket: this.bucket,
        Key: key,
      }));

      return {
        exists: true,
        size: result.ContentLength,
        lastModified: result.LastModified,
        contentType: result.ContentType,
      };
    } catch (error: any) {
      if (error.name === "NotFound" || error.$metadata?.httpStatusCode === 404) {
        return { exists: false };
      }
      
      console.error("Failed to get file info:", error);
      throw new Error(`Failed to get file info: ${error.message}`);
    }
  }
}

// Export default instance
export const s3Upload = new S3MultipartUpload();

// Utility function to validate AWS configuration
export function validateAWSConfig(): { isValid: boolean; missing: string[] } {
  const required = [
    "AWS_ACCESS_KEY_ID",
    "AWS_SECRET_ACCESS_KEY", 
    "AWS_S3_BUCKET",
  ];
  
  const missing = required.filter(key => !process.env[key]);
  
  return {
    isValid: missing.length === 0,
    missing,
  };
} 